crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -46,10 +46,23 @@ impl<'db> UnionBuilder<'db> {
     pub(crate) fn add(mut self, ty: Type<'db>) -> Self {
         match ty {
             Type::Union(union) => {
-                self.elements.extend(union.elements(self.db));
+                for element in union.elements(self.db) {
+                    self = self.add(*element);
+                }
             }
             Type::Never => {}
             _ => {
+                let mut remove = vec![];
```

Comment:
In equivalent mypy code, I had to add a special fast path for literals. You can do better than n**2 for unions with lots of literals of the same type, which turns out to be a thing in the wild

Reply:
Interesting, thanks for pointing this out! I took a look at that optimization in mypy.

I think the case this would optimize is where a union already contains e.g. `str`, and then we try to add lots of string literal types to it, every one of which is redundant because its a subtype of `str`. Rather than going through all existing union members to check if each literal is a subtype of any of them, we can keep a hash-set of "types that have literal forms which are present in this union" and do an O(1) contains check against that set as the first step when adding a literal type to the union. Framed in more general terms, it's identifying that a certain set of common types have a single super-type that is most likely to rule them out of the union, and so we optimize checking for that most likely super-type by identity.

This makes sense; I'd prefer to wait to add this kind of optimization until we see it crop up in a real-world codebase and can evaluate the actual impact of the optimization in our case, but it's definitely a useful idea to keep in mind.

Reply:
I think what would be useful is if we added one (or more) benchmarks based on a real-world codebase that makes heavy use of large literals. (I.e., pydantic.)

Reply:
That's not quite the right description, it's also useful when the union doesn't contain the supertype (i.e. str). For instance, say if you were combining two unions that you knew consisted only of literal types, you could use a set union, which is linear. The mypy optimisation I added is basically that, but also works when there are non-literal types thrown in as well. Fair enough on waiting though!

Reply:
Oh, thanks, yeah, I misread the code. The set is `unduplicated_literal_fallbacks`, not `duplicated_literal_fallbacks`. So it looks like it's optimizing only the case you described; the mirror image of the case I described.

Reply:
I've created a new issue collating some of the perf issues mypy and pyright have encountered relating to unions: https://github.com/astral-sh/ruff/issues/13549

Reply:
Nice!
I think there are some mypy PRs missing from the list, so if you're interested in code I'd make sure to look at main.
I'll also make it such that if you're interested in real world use cases you should only have to look at primer.